src/shared/lib/html-table-validator.js

```diff
@@ -0,0 +1,289 @@
+* unknown or incorrect rules will throw an exception.
+* we don't check column or row umber validity: if you put "column: 'a'",
+  that's on you!
+
+With the rules described, instatiate a validator, and give it your
+cheerio table:
+
+    const v = new HtmlTableValidator(rules);
+    if (!v.success($table)) {
+      // Some rules failed ...
+      console.log(v.errors($table);
+    }
+
+Or as a shorthand, just throw and optionally log to console:
+
+    const opts = { includeErrCount: 5, logToConsole: false };
+    HtmlTableValidator.throwIfErrors($rules, $table, opts);
+
+
+Usage notes:
+
+There are some ASSUMPTIONS documented in the code:
+
+* headings are on row 0
+* data starts on row 1
+* one data point per table cell
+
+*/
+class HtmlTableValidator {
+  constructor(rules) {
+    const setrules = {
+      headings: {},
+      minrows: 0,
+      data: []
+    };
+
+    for (const k in rules) {
+      if (!Object.keys(setrules).includes(k)) {
+        const msg = `bad rule key ${k}`;
+        // console.log(msg);
+        throw new Error(msg);
+      }
+    }
+
+    // eslint-disable-next-line guard-for-in
+    for (const k in rules) setrules[k] = rules[k];
+
+    this._validateRules(setrules);
+    this.rules = setrules;
+  }
+
+  // PUBLIC //////////////////////////////////////////////
+
+  success(table) {
+    return this.errors(table).length === 0;
+  }
+
+  // Returns list of error messages.
+  errors(table) {
+    const result = [];
+    if (table === null || table === undefined) {
+      result.push('null/undefined table');
+      return result;
+    }
+
+    // ASSUMPTION: table must have 1 header row,
+    // and at least one data row.
+    const trs = table.find('tr');
+    if (trs.length <= 1) {
+      return ['no rows in table'];
+    }
+
+    result.push(...this._checkHeadings(table, this.rules.headings));
+    result.push(...this._checkMinRows(table, this.rules.minrows));
+    result.push(...this._checkData(table, this.rules.data));
+
+    return result;
+  }
+
+  // Utility method to do quick validation.
+  //
+  // Options defaults:
+  // {
+  //   includeErrCount: 5,
+  //   logToConsole: false
+  // }
+  static throwIfErrors(rules, table, options = {}) {
+    const includeErrCount = options.includeErrCount || 5;
+    const logToConsole = options.logToConsole || false;
+
+    const v = new HtmlTableValidator(rules);
+    const errs = v.errors(table);
+    const errCount = errs.length;
+    if (errCount === 0) return;
+
+    const msg = `${errCount} validation errors.`;
+    const firstN = errs.slice(0, includeErrCount);
+    if (logToConsole) {
+      console.error(msg);
+      console.error(firstN);
+    }
+    throw new Error(`${msg}.  Sample: ${firstN.join(';')}.`);
+  }
+
+  // PRIVATE (ish) //////////////////////////////////////////////
+  //
+  // Javascript doesn't really have private, prepending underscore
+  // to discourage direct calls.
+  //
+  // Most of these methods should be static b/c they don't use "this",
```

Comment:
It's not really "broken" :-) , but I wasn't sure how to handle this.  These are helper functions only.  I found it code clutter to say `HtmlTableValidator.my_func()` instead of `my_func()`, but the linter insisted on the longer name.  Since this was lint only, I figured hiding the warning was ok.

ps - This is me not knowing javascript, if there's a better way to do it, LMK!